### PR TITLE
Cache optimizations 

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Increase speed by more optimized cache maintenance. Also some TLB invalidations are optimized.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
On Hikey running xtest with paging enabled in AArch32:
Before:
```
time xtest
real    1m 38.05s
user    0m 0.08s
sys     1m 30.69s
```

after:
```
time xtest
real    1m 29.48s
user    0m 0.12s
sys     1m 18.83s

**Comment by jforissier** [comment]
I get a crash on HiKey960 (64-bit TAs), which seems to be triggered by commit 227c51ae587c ("core: pager: use icache_inv_user_range()"). At least that's what "git bisect" says.

```
3.5.0-76-g20d95107
make -j10 CFG_ULIBS_SHARED=y CFG_WITH_PAGER=y COMPILE_S_USER=64 CFG_TEE_CORE_LOG_LEVEL=2

$ xtest 6016
[...]
o regression_6016.2 Storage id: 80000000
    threads: 4, loops: 8
E/TC:5 3
E/TC:5 3 User TA data-abort at address 0x7f48 (translation fault)
E/TC:5 3  esr 0x92000005  ttbr0 0x600003f0240a0   ttbr1 0x00000000   cidr 0x0
E/TC:5 3  cpu #5          cpsr 0x60000100
E/TC:5 3  x0  ffffffffffffffe0 x1  0000000000007f40
E/TC:5 3  x2  0000000000007f60 x3  0000000040030570
E/TC:5 3  x4  0000000040030560 x5  000000004002d3c8
E/TC:5 3  x6  0000000000005135 x7  0000000000000020
E/TC:5 3  x8  0000000040005b40 x9  0000000000005135
E/TC:5 3  x10 000000004004d980 x11 0000000000000000
E/TC:5 3  x12 0000000000000000 x13 0000000040025f80
E/TC:5 3  x14 0000000000000000 x15 0000000000000000
E/TC:5 3  x16 0000000040046098 x17 0000000040032630
E/TC:5 3  x18 0000000000000000 x19 0000000000000000
E/TC:5 3  x20 0000000000000004 x21 0000000040025f88
E/TC:5 3  x22 0000000000005135 x23 0000000040025f80
E/TC:5 3  x24 0000000000000004 x25 000000003f036720
E/TC:5 3  x26 000000003f08bf00 x27 0000000000000004
E/TC:5 3  x28 0000000040025f88 x29 0000000040025e80
E/TC:5 3  x30 00000000400327d8 elr 0000000040032834
E/TC:5 3  sp_el0 0000000040025e80
E/TC:5 3 Status of TA b689f2a7-8adf-477a-9f99-32e90c0ad0a2 (0x3f036940) (active)
E/TC:5 3  arch: aarch64  load address: 0x0000000040026000 ctx-idr: 6
E/TC:5 3  stack: 0x0000000040025000 4096
E/TC:5 3  region  0: va 0x0000000040000000 pa 0x000000003f002000 size 0x002000 flags ---R-X
E/TC:5 3  region  1: va 0x0000000040002000 pa 0x000000003f016000 size 0x001000 flags ---RW-
E/TC:5 3  region  2: va 0x0000000040004000 pa 0x00000000ba489b20 size 0x001000 flags rw-RW-
E/TC:5 3  region  3: va 0x0000000040005000 pa 0x00000000ba489b40 size 0x001000 flags rw-RW-
E/TC:5 3  region  4: va 0x0000000040025000 pa 0x0000000000000000 size 0x001000 flags rw-RW-
E/TC:5 3  region  5: va 0x0000000040026000 pa 0x0000000000000000 size 0x002000 flags r-xR-- [0]
E/TC:5 3  region  6: va 0x0000000040028000 pa 0x0000000000000000 size 0x009000 flags rw-RW- [0]
E/TC:5 3  region  7: va 0x0000000040031000 pa 0x0000000000000000 size 0x005000 flags r-xR-- [1]
E/TC:5 3  region  8: va 0x0000000040045000 pa 0x0000000000000000 size 0x002000 flags rw-RW- [1]
E/TC:5 3  region  9: va 0x0000000040047000 pa 0x0000000000000000 size 0x00f000 flags r-xR-- [2]
E/TC:5 3  region 10: va 0x0000000040065000 pa 0x0000000000000000 size 0x004000 flags rw-RW- [2]
E/TC:5 3  region 11: va 0x0000000040069000 pa 0x0000000000000000 size 0x036000 flags r-xR-- [3]
E/TC:5 3  region 12: va 0x00000000400ae000 pa 0x0000000000000000 size 0x005000 flags rw-RW- [3]
E/TC:5 3  [0] b689f2a7-8adf-477a-9f99-32e90c0ad0a2 @ 0x0000000040026000
E/TC:5 3  [1] 71855bba-6055-4293-a63f-b0963a737360 @ 0x0000000040031000
E/TC:5 3  [2] 527f1a47-b92c-4a74-95bd-72f19f4a6f74 @ 0x0000000040047000
E/TC:5 3  [3] 87bb6ae8-4b1d-49fe-9986-2b966132c309 @ 0x0000000040069000
regression_6000.c:1700: fs_create(&sess, a->file_name, sizeof(a->file_name), 0x00000002, 0, data_01, sizeof(data_01), &obj, a->storage_id) has an unexpected value: 0xffff3024 = TEE_ERROR_TARGET_DEAD, expected 0x0 = TEEC_SUCCESS
  regression_6016.2 FAILED
  regression_6016 FAILED
```

**Comment by jenswi-linaro** [comment]
Thanks, I'll try to stress my Hikey a bit more.

**Comment by jforissier** [comment]
FYI, sometimes you need 2 or 3 runs of `xtest 6016` to trigger the crash.

**Comment by etienne-lms** [comment]
Successfully tested on armv7 (stm32mp1, with LPAE. Found #3009 but that unrelated.
I will run some stress tests with and wothout LPAE.

**Comment by jenswi-linaro** [comment]
Rebased on master and also added a fix that hopefully fixes the issue found by @jforissier.

**Comment by jenswi-linaro** [comment]
@etienne-lms, what kind of i-cache does stm32mp1 have? PIPT or VIPT?

**Comment by etienne-lms** [comment]
L1 icache is VIPT on stm32mp1.

**Comment by jenswi-linaro** [comment]
Thanks. I guess the reason why you didn't see the error was that you didn't compile with `CFG_ULIBS_SHARED=y CFG_WITH_PAGER=y CFG_TEE_CORE_LOG_LEVEL=2`.

**Comment by jforissier** [comment]
@jenswi-linaro yes, the issue I mentioned in https://github.com/OP-TEE/optee_os/pull/3005#issuecomment-492130261 cannot be reproduced now (commit 76525ef).

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey960)`

(I have tested pretty much all the configurations I run for release testing; commit 76525ef4e7f1 + PR 3086 on top).

**Comment by jenswi-linaro** [comment]
Thanks for the rigorous testing. The code seems to work OK, now it only need some reviewing too.

File: core/arch/arm/include/arm64.h [hunk 1]
```diff
@@ -249,6 +249,11 @@ static inline void tlbi_vaae1is(uint64_t mva)
 	asm volatile ("tlbi	vaae1is, %0" : : "r" (mva));
 }
 
+static inline void tlbi_vale1is(uint64_t mva)
+{
+	asm volatile ("tlbi	vaae1is, %0" : : "r" (mva));
```
Comment (jforissier) [inline core/arch/arm/include/arm64.h]: The name of the function is `*val*` but here we have `vaa`, why? Spelling mistake?

**Comment by jforissier** [comment]
For commits:
- "arm.h: add CTR_WORD_SIZE"
- "core: cache_helpers_a{32,64}.S: remove section assignments"
- "core: arm32.h: add TLBI_{MVA_SHIFT,ASID_MASK}"

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/arch/arm/include/arm64.h (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/include/arm64.h]: Well spotted, it's a spelling mistake. I'll fix.

**Comment by jenswi-linaro** [comment]
Squashed, rebased and tags applied. Note that a bunch of commits are missing R-B or A-B tags.

File: core/arch/arm/kernel/thread_a64.S [hunk 2]
```diff
@@ -703,6 +703,92 @@ eret_to_el0:
 #endif /*CFG_CORE_UNMAP_CORE_AT_EL0*/
 
 	eret
+
+	/*
+	 * void icache_inv_user_range(void *addr, size_t size);
+	 *
+	 * This function has to execute with the user space ASID active,
+	 * this means executing with reduced mapping and the code need
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_a64.S]: s/need/needs/

File: core/arch/arm/kernel/thread_a64.S [hunk 3]
```diff
@@ -703,6 +703,92 @@ eret_to_el0:
...
+#endif /*CFG_CORE_UNMAP_CORE_AT_EL0*/
+
+	/*
+	 * Do the actual icache invalidation
+	 */
+
+	/* Calculate minimum icache line size, result in x2 */
+	mrs	x3, ctr_el0
+	and	x3, x3, #CTR_IMINLINE_MASK
+	mov	x2, #4
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_a64.S]: s/#4/#CTR_WORD_SIZE/

File: core/arch/arm/mm/tee_pager.c [hunk 4]
```diff
@@ -1012,22 +1014,55 @@ static bool tee_pager_unhide_page(struct tee_pager_area *area,
...
+	 * The page is hidden, or not not mapped yet. Unhide the page and
+	 * move it to the tail.
+	 *
+	 * Since the page isn't mapped there doesn't exist a valid TLB entry
+	 * for this address, so no TLB invalidation is required after setting
+	 * the new entry. A DSB is needed though, to make the write visible.
+	 *
+	 * For user executable pages it's more complicated. Those pages can
+	 * be shared between multiple TA mappings and thus populated by
+	 * another TA. The reference manual statates that:
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c]: typo: s/statates/states/

**Comment by jenswi-linaro** [comment]
Comments address, commits slightly reordered with some conflict resolving and tags applied.

**Pull request merged**